File: c/sort/array/MergeSort.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static void sort(int *nums, int *aux, int start, int end);
static void merge(int *nums, int *aux, int start, int middle, int end);

void mergeSort(int *nums, int length) {
    int *aux = (int *) malloc(sizeof(int) * length);
    memcpy(aux, nums, sizeof(int) * length);
    sort(nums, aux, 0, length - 1);
}

static void sort(int *nums, int *aux, int start, int end) {
    if (start >= end) {
        return;
    }

    int middle = start + ((end - start) >> 1);
    sort(aux, nums, start, middle);
    sort(aux, nums, middle + 1, end);

    if (aux[middle] > aux[middle + 1]) {
        merge(nums, aux, start, middle, end);
    } else {
        memcpy(nums + start, aux + start, sizeof(int) * (end - start + 1));
    }
}

static void merge(int *nums, int *aux, int start, int middle, int end) {
    int index = start;
    int left = start;
    int right = middle + 1;
    while (left <= middle && right <= end) {
        if (aux[left] > aux[right]) {
            nums[index] = aux[right];
            ++right;
        } else {
            nums[index] = aux[left];
            ++left;
        }
        ++index;
    }

    if (left <= middle) {
        memcpy(nums + index, aux + left, sizeof(int) * (middle - left + 1));
    }
    if (right <= end) {
        memcpy(nums + index, aux + right, sizeof(int) * (end - right + 1));
    }
}

File: py/sort/linked_list/MergeSort.py
#!/usr/bin/env python
# -*- coding:gbk -*-

class ListNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class MergeSort:
    def sort(self, head: ListNode) -> None:
        self.__sort(head, head.next)

    def __sort(self, head: ListNode, start: ListNode) -> None:
        if not start.next:
            head.next = start
            return

        left, right, middle = ListNode(0), ListNode(0), self.__getMiddle(start)
        left.next, right.next, middle.next = start, middle.next, None

        self.__sort(left, left.next)
        self.__sort(right, right.next)
        self.__merge(head, left.next, right.next)

    def __merge(self, head: ListNode, left: ListNode, right: ListNode) -> None:
        current = head
        while left and right:
            if left.value > right.value:
                current.next, right = right, right.next
            else:
                current.next, left = left, left.next
            current = current.next

        if left:
            current.next = left
        if right:
            current.next = right

    def __getMiddle(self, start: ListNode) -> ListNode:
        if not start.next:
            return start

        slow, fast = start, start.next
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next
        return slow
